### PR TITLE
RTC cleanup, documentation, setting date/time

### PR DESCRIPTION
**Pull request by gyrovorbis**
I have been referencing the RTC a bunch with the date/time work for Newlib and now the NS timer stuff with the perf counters that I'm working on... I noticed the docs could be fleshed out better and there was no way to SET the date/time. 

- Improved Doxygen documentation for RTC, adding more information on it and a new module
- Added support for setting the date/time
- Removed magic numbers
- Added a #define for configuring retry attempts for read/write failures
- Added explicit `(void)` argument lists for type checking parameters properly in plain C
- Tested on my own DC, ensuring that the cached boot time was updated properly as well and that standard C date/time functions behave properly with change

### CODE REVIEW

File: kernel/arch/dreamcast/kernel/rtc.c [hunk 1]
```diff
@@ -22,13 +25,20 @@
  */
 
 #include <arch/rtc.h>
+#include <arch/timer.h>
 #include <dc/g2bus.h>
 
+#define RTC_CONTROL_ADDR        0xa0710008  /* SFR with write-enable bit */
+#define RTC_SECONDS_HIGH_ADDR   0xa0710000  /* Upper 16 bits of timestamp */
+#define RTC_SECONDS_LOW_ADDR    0xa0710004  /* Lower 16 bits of timestamp */
```
Comment (QuzarDC) [inline kernel/arch/dreamcast/kernel/rtc.c:33]: I'd suggest that the three register values be in address order (CTRL last), and be moved into the header file. With them in the header they can get doxy comments added, which would help tracking and hw documentation. Then could also note the behavior that is in comments of the set time function with the CTRL being write-only and self clearing when the high_addr is written to.

Though it's possibly overly pedantic, I might also have '#define RTC_CTRL_ADDR_WRITE_ENABLE (1 << 0)' or similar.
Or having RTC_CONTROL_ADDR be RTC_WRITE_ENABLE then define the value as RTC_WRITE_ENABLE_SET. In any case tagging in #151 for visibility there.

File: kernel/arch/dreamcast/kernel/rtc.c [hunk 2]
```diff
@@ -52,23 +63,66 @@ time_t rtc_unix_secs() {
...
+
+    /* Try 3 times to ensure we didn't write a value then have 
+       the clock increment itself before the next. */
+    for(i = 0; i < RTC_RETRY_COUNT; i++) { 
+        /* Write the least-significant 16-bits first, because 
+           writing to the high 16-bits will lock RTC writes. */
+        g2_write_32(RTC_SECONDS_LOW_ADDR, (adjusted) & 0xffff);
+        g2_write_32(RTC_SECONDS_HIGH_ADDR, (adjusted >> 16) & 0xffff);
+
+        /* Read the time back again, to ensure it was written properly. */
```
Comment (QuzarDC) [inline kernel/arch/dreamcast/kernel/rtc.c:92]: Should this just call rtc_unix_secs? If the reading is fragile, then couldn't it be possible that the write succeeds properly, but the read to verify it is broken without doing the looping? Then we loop over and do 2 invalid writes and maybe still not validate?

If the writing is fragile, as the comments suggest, I could also see a situation where on the last retry the write succeeds then we only get one chance to read correctly.

File: kernel/arch/dreamcast/include/arch/rtc.h [hunk 3]
```diff
@@ -23,29 +27,78 @@ __BEGIN_DECLS
...
 #include <time.h>
 
-/** \brief  Get the current date/time.
+/** \defgroup rtc Real-Time Clock
+    \brief Real-Time Clock Management 
+
+    Provides an API for fetching and managing the date/time using
+    the Dreamcast's real-time clock. All timestamps are in standard
+    Unix format, with an epoch of January 1, 1970. Due to the fact
+    that there is no time zone data on the RTC, all times are expected
```
Comment (QuzarDC) [inline kernel/arch/dreamcast/include/arch/rtc.h:36]: Not sure if it should be here or in the brief, but there should be at least once place where RTC is glossed. So "Real-Time Clock (RTC) Management"

**Review by QuzarDC — COMMENTED** [review]
Overall approve, but the looping in the set seems off to me (as noted in my comments).

File: kernel/arch/dreamcast/kernel/rtc.c (hunk 1 above)
Comment (gyrovorbis) [reply kernel/arch/dreamcast/kernel/rtc.c:33]: I had actually considered moving the registers to the header file, but was worried you guys might think that was beyond the scope of what this driver should be doing...

BUT I ACTUALLY TOTALLY AGREE! That's a fantastic idea. I think every header should document the hardware publicly as well, like that for our KOS hardware documentation in doxygen!

Agreed on what you're saying about having a define for the actual write enable bit. I'll get on this.

File: kernel/arch/dreamcast/include/arch/rtc.h (hunk 3 above)
Comment (gyrovorbis) [reply kernel/arch/dreamcast/include/arch/rtc.h:36]: Good call!

**Review by ljsebald — APPROVED** [review]

**Pull request merged**